src/consensus/proposer.rs

```diff
@@ -73,28 +73,34 @@ impl ProposedValues {
 
     fn add_proposed_value(&mut self, value: FullProposal) {
         let height = value.height();
-        let round = value.round();
         let shard_hash = value.shard_hash();
         self.values.insert(shard_hash.clone(), value);
-        self.values_by_height.insert((height, round), shard_hash);
+        match self.values_by_height.get_mut(&height) {
+            Some(hashes) => {
+                hashes.push(shard_hash);
+            }
+            None => {
+                self.values_by_height.insert(height, vec![shard_hash]);
+            }
+        }
     }
 
     fn get_by_shard_hash(&self, shard_hash: &ShardHash) -> Option<&FullProposal> {
         self.values.get(&shard_hash)
     }
 
-    fn decide(&mut self, height: Height, round: Round) {
-        while let Some(((entry_height, entry_round), entry_shard_hash)) =
-            self.values_by_height.pop_first()
-        {
-            if entry_height < height || (entry_height == height && entry_round < round) {
-                self.values.remove(&entry_shard_hash);
+    fn decide(&mut self, height: Height) {
+        while let Some((entry_height, entry_shard_hashes)) = self.values_by_height.pop_first() {
+            if entry_height <= height {
+                for entry_shard_hash in entry_shard_hashes {
+                    self.values.remove(&entry_shard_hash);
+                }
                 continue;
             }
 
             // Put it back in, we shouldn't have removed
             self.values_by_height
-                .insert((entry_height, entry_round), entry_shard_hash);
+                .insert(entry_height, entry_shard_hashes);
```

Comment:
Hmm, this insert feels dangerous. There's a potential for a race condition here where we've added a value for a height and then this insert replaces it with an outdated vector.

I think we should iterate and collect the heights we want to remove first and then remove it in one go or always remove only the previous height if it exists. The latter seems cleaner?

Reply:
But I think it's more correct to remove all previous heights. So, would be better to do the former